src/relay.cpp

```diff
@@ -0,0 +1,196 @@
+// Copyright 2021 Mateusz Lichota
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <rclcpp/generic_publisher.hpp>
+#include <rclcpp/generic_subscription.hpp>
+
+#include <memory>
+#include <string>
+
+using namespace std::chrono_literals;
+
+const auto LOOP_PERIOD = 100ms;
+
+namespace topic_tools
+{
+
+  class RelayNode : public rclcpp::Node
+  {
+  public:
+    RelayNode(
+      const std::string node_name, 
+      const rclcpp::NodeOptions &options);
+      
+    RelayNode(
+      const std::string node_name, 
+      const std::string input_topic, 
+      const std::string output_topic);
+
+  private:
+    void incoming_message_callback(std::shared_ptr<rclcpp::SerializedMessage> msg);
+    void timer_callback();
+    void subscribe();
+    void unsubscribe();
+    void initialize();
+    std::optional<std::string> try_find_topic_type();
+
+    rclcpp::GenericSubscription::SharedPtr sub_;
+    rclcpp::GenericPublisher::SharedPtr pub_;
+    rclcpp::TimerBase::SharedPtr timer_;
+    std::string input_topic;
+    std::string output_topic;
+    bool lazy;
+    std::optional<std::string> topic_type;
+  };
+
+  RelayNode::RelayNode(
+    const std::string node_name, const rclcpp::NodeOptions &options = rclcpp::NodeOptions()) 
+    : rclcpp::Node(node_name, options)
+  {
+    input_topic = declare_parameter<std::string>("input_topic");
+    output_topic = declare_parameter<std::string>("output_topic", input_topic + "_relay");
+    initialize();
+  }
+
+  RelayNode::RelayNode(
+      std::string node_name,
+      std::string input_topic,
+      std::string output_topic) : rclcpp::Node(node_name),
+                                  input_topic(input_topic),
+                                  output_topic(output_topic)
+  {
+    initialize();
+  }
+
+  void RelayNode::initialize()
+  {
+    lazy = declare_parameter<bool>("lazy", false);
+
+    auto ros_clock = rclcpp::Clock::make_shared();
+    timer_ = rclcpp::create_timer(this, ros_clock, LOOP_PERIOD, [=]()
+                                  { timer_callback(); });
```

Comment:
you won't need to construct your own clock or call global functions - I would recommend using the `Node::create_wall_timer` function instead

```suggestion
    discovery_timer_ = create_timer(LOOP_PERIOD, std::bind(RelayNode::discovery_topic_types, this));
```